Bug fixes and added more tests

lnedry opened the pull request:
Changes proposed in this pull request:

- check host_list to confirm if local domain
- added config file for allowed Message-IDs
- added tests for bounce_spf()
- added Sinon for creating mocks
- moved bad_rcpt function from hook_data to hook_rcpt_ok
- renamed config key check.reject_all to reject.all_bounces
- moved config key check.bad_rcpt to reject.bad_rcpt
- refactored most of the existing tests to be more thorough
- Message-ID header regex to check for required angle brackets
- moved reject_all from [check] to [reject] and renamed to all_bounces

Fixes #

- fixed bug where config key reject.empty_return_path was ignored
- fixed timeout bug in bounce_spf()
- updated bounce_spf to use promises

Checklist:

- [ ] docs updated
- [X] tests updated
- [ ] Changes.md updated
- [ ] package.json.version bumped

lnedry committed:
Update README.md

Update empty_return_path to match default in source code.

lnedry committed:
Update README.md

Moved reject_all=false from [check] to [reject] and renamed to all_bounces=false

lnedry committed:
Update bounce.ini

Moved reject_all from [check] to [reject] and renamed to all_bounces

lnedry committed:
Update package.json

Added Sinon to create mocks for testing

lnedry committed:
Update index.js

- Updated bounce_spf() to use Promises.
- Moved bad_rcpt function from hook_data to hook_rcpt_ok.
- Renamed config key check.reject_all to reject.all_bounces.
- Moved config key check.bad_rcpt to reject.bad_rcpt
- Fixed bug where config key reject.empty_return_path was ignored.
- Fixed timeout bug in bounce_spf()
- Modified Message-ID header regex to check for required angle brackets.

lnedry committed:
Update bounce.ini

Moved config key check.bad_rcpt to reject.bad_rcpt.

lnedry committed:
Update index.js

- Added tests for bounce_spf()
- Refactored most of the existing tests to be more thorough.

lnedry committed:
Update README.md

Moved config key check.bad_rcpt to reject.bad_rcpt

msimerson commented on index.js:
```diff
@@ -51,10 +64,15 @@ exports.load_bounce_ini = function () {
+  // legacy settings
+  const c = this.cfg
+  if (c.check.reject_all) c.reject.all_bounces = c.check.reject_all
+  if (c.check.reject) c.reject.reject = c.check.reject
```
this doesn't look right

msimerson commented on index.js:
```diff
@@ -103,16 +121,19 @@ exports.single_recipient = function (next, connection) {
+  }
 
-  next(DENY, 'this bounce message does not have 1 recipient')
+  return next()
```
```suggestion
  next()
```

Avoid a useless use of `return` here.

msimerson commented on index.js:
```diff
@@ -128,40 +149,35 @@ exports.empty_return_path = function (next, connection) {
 
-  transaction.results.add(this, { fail: 'empty_return_path', emit: true })
-  return next(DENY, 'bounce with non-empty Return-Path (RFC 3834)')
+  return next()
```
```suggestion
  next()
```

Again, avoid useless use of `return`.

msimerson commented on index.js:
```diff
@@ -128,40 +149,35 @@ exports.empty_return_path = function (next, connection) {
+  connection.transaction.results.add(this, { pass: 'bad_rcpt' })
 
-  next()
+  return next()
```
```suggestion
  next()
```

avoid useless use of return. (The calling function does nothing with the functions return. Not using `return next...` everywhere makes it subtly more obvious that the `return next...` patterns above are *strictly* for flow control.)

msimerson commented on index.js:
```diff
@@ -128,40 +149,35 @@ exports.empty_return_path = function (next, connection) {
 exports.has_null_sender = function (connection, mail_from) {
-  const transaction = connection?.transaction
-  if (!transaction) return false
+  const { transaction } = connection
```
You're no longer guarding against the transaction having been destroyed. How can you be certain that when this function is called, the transaction still exists?

lnedry replied:
Under what circumstances could the transaction be destroyed before one of these hooks gets called?

msimerson replied:
It's most common when the remote disconnects in the middle of the conversation. Any hooks that were running *before* the connection dropped will still be running. The transaction may get cleaned up while a plugin hook is running, especially if the hook does anything asynchronous. Then subsequent attempts to access `connection.transaction` will be undefined and trigger an exception, which takes down the node process.

msimerson commented on index.js:
```diff
@@ -172,29 +188,27 @@ exports.has_null_sender = function (connection, mail_from) {
 
-  const transaction = connection?.transaction
-  if (!transaction) return next()
+  const { transaction } = connection
```
same here, I'm not sure we want to remove the transaction guarding.

msimerson replied:
```suggestion
  const { transaction } = connection
  if (!transaction) return
```

msimerson commented on index.js:
```diff
@@ -205,71 +219,69 @@ exports.non_local_msgid = function (next, connection) {
-  //         {fail: 'Message-ID not local', emit: true });
-  // if (!plugin.cfg.reject.non_local_msgid) return next();
-  // return next(DENY, "bounce with non-local Message-ID (RFC 3834)");
+  return next()
```
```suggestion
  next()
```

msimerson commented on index.js:
```diff
@@ -319,74 +329,50 @@ exports.bounce_spf = function (next, connection) {
+    return next(DENY, 'Invalid bounce (spoofed sender)')
+  }
+
+  return next()
```
```suggestion
  next()
```

msimerson reviewed (commented):
I like this PR, especially the improvement of config naming.

lnedry committed:
Update index.js

Removed spurious line of code

lnedry committed:
Update README.md

- Removed description for reject_all
- Added description for all_bounces

lnedry committed:
Update index.js

Removed useless use of return

lnedry committed:
Update index.js

Removed useless use of return

lnedry committed:
Update index.js

Restored transaction guarding

lnedry committed:
Update index.js

Added test to check for missing transaction

lnedry commented:
Removed spurious line of code.
Changed several return next() statements to just next()
Restored transaction guarding.
Added test for transaction guarding.
Replaced description in README for reject_all with description for all_bounces

msimerson commented on index.js:
```diff
@@ -103,16 +120,19 @@ exports.single_recipient = function (next, connection) {
 
-  if (!this.cfg.reject.single_recipient) return next()
+  if (this.cfg.reject.single_recipient) {
+    next(DENY, 'this bounce message has too many recipients')
```
```suggestion
    return next(DENY, 'this bounce message has too many recipients')
```

You do need the return here, for flow control. Otherwise you'll be calling `next()` twice.

msimerson commented on index.js:
```diff
@@ -128,40 +148,36 @@ exports.empty_return_path = function (next, connection) {
-    const rcpt = element.address()
-    if (!this.cfg.invalid_addrs[rcpt]) continue
-    transaction.results.add(this, { fail: 'bad_rcpt', emit: true })
+  if (this.cfg.invalid_addrs.includes(rcpt.address())) {
```
Is the result of `rcpt.address()` always lower case?

msimerson commented on index.js:
```diff
@@ -128,40 +148,36 @@ exports.empty_return_path = function (next, connection) {
-  const transaction = connection?.transaction
-  if (!transaction) return false
+  const { transaction } = connection
+  if (!transaction) return next()
```
```suggestion
  if (!transaction) return false
```

There's no `next()` in context here.

msimerson commented:
Perhaps a good test to add for each bounce test is passing in a connection without a transaction, and assuring that the function returns (doesn't trigger an exception).

lnedry committed:
Update index.js

minor cose cleanup

msimerson commented on index.js:
```diff
@@ -293,24 +305,22 @@ function find_received_headers(ips, body, connection, self) {
 }
 
 exports.bounce_spf_enable = function (next, connection) {
-  if (!connection.transaction) return next()
```
restore transaction guarding here.

msimerson commented on index.js:
```diff
@@ -293,24 +305,22 @@ function find_received_headers(ips, body, connection, self) {
 
-  const txn = connection?.transaction
-  if (!txn) return next()
+  const { transaction } = connection
```
```suggestion
  const { transaction } = connection
  if (!transaction) return next()
```

msimerson reviewed (changes requested):
These are the last few issues I see.

lnedry committed:
Update index.js

- added tests for transaction guarding
- moved transaction guarding from has_null_sender() to each hook
- refactored has_null_sender()

lnedry committed:
Update index.js

Added tests for guarding transactions

msimerson reviewed (changes requested):
Looking good! 

A final issue is documentation. You've added the use of three new config files to the plugin:

  - host_list
  - bounce_bad_rcpt
  - bounce_allowed_msgids

None of those files are mentioned in the README, nor is the format for what you'd put in them mentioned. There are alternatives to accomplish this:

## 1. Move the data into bounce.ini and add inline docs there

```ini

[check]
single_recipient=true
empty_return_path=true

[reject]
single_recipient=true
empty_return_path=true

; reject all bounce messages (generally a bad idea)
all_bounces=false

; reject bounces to "bad" recipients
bad_rcpt=true

; bad_rcpts[] = nobody@example.com
; bad_rcpts[] = no_reply@example.com
; bad_rcpts[] = ....


; to validate bounce message IDs, add an entry to the domains array
; for each email domain you accept email for.
[msgid]
; domains[] = example.com
; domains[] = example.net
; domains[] = example.org
```

## 2. Add a section to the README for each config file

Explain why you'd use the file and what the contents of the file should be.

Finally, the local msgid checks the domain against the contents of the host_list file. In my case, there's 4 domains listed in hosts_list, but that file hasn't been updated since 2021 and the rcpt_to.in_hosts_list plugin isn't even loaded. I think you want to say something like this:

- The Message-ID test needs to know what domains are 'local' to your mail server. If you use the rcpt_to.host_list plugin and have populated the `config/host_list` file, then you're done.
- If you don't use that plugin, make sure the host_list file is empty and populate the [msgid].domains[] array in bounce.ini (or config/bounce_allowed_msgids).

Also, if you keep the file name bounce_allowed_msgids, a name that better indicates what goes in it might be bounce_msgid_allowed_domains?

lnedry committed:
Update bounce.ini

updated settings to match defaults in source

lnedry committed:
Update README.md

- updated documentation for all_bounces and bad_rcpt
- added documentation for non_local_msgid

lnedry committed:
Update index.js

minor code clean up

lnedry committed:
Update index.js

cleaned up test code
